x/mongo/driver/topology/topology_options.go

```diff
@@ -7,451 +7,345 @@
-	data, err := ioutil.ReadFile(clientFile)
-	if err != nil {
-		return "", err
+	// TLSConfig
+	if co.TLSConfig != nil {
+		connOpts = append(connOpts, WithTLSConfig(
+			func(*tls.Config) *tls.Config {
+				return co.TLSConfig
+			},
+		))
 	}
 
-	var currentBlock *pem.Block
-	var certBlock, certDecodedBlock, keyBlock []byte
-
-	remaining := data
-	start := 0
-	for {
-		currentBlock, remaining = pem.Decode(remaining)
-		if currentBlock == nil {
-			break
-		}
-
-		if currentBlock.Type == "CERTIFICATE" {
-			certBlock = data[start : len(data)-len(remaining)]
-			certDecodedBlock = currentBlock.Bytes
-			start += len(certBlock)
-		} else if strings.HasSuffix(currentBlock.Type, "PRIVATE KEY") {
-			if keyPasswd != "" && x509.IsEncryptedPEMBlock(currentBlock) {
-				var encoded bytes.Buffer
-				buf, err := x509.DecryptPEMBlock(currentBlock, []byte(keyPasswd))
-				if err != nil {
-					return "", err
-				}
-
-				pem.Encode(&encoded, &pem.Block{Type: currentBlock.Type, Bytes: buf})
-				keyBlock = encoded.Bytes()
-				start = len(data) - len(remaining)
-			} else {
-				keyBlock = data[start : len(data)-len(remaining)]
-				start += len(keyBlock)
-			}
-		}
-	}
-	if len(certBlock) == 0 {
-		return "", fmt.Errorf("failed to find CERTIFICATE")
-	}
-	if len(keyBlock) == 0 {
-		return "", fmt.Errorf("failed to find PRIVATE KEY")
+	// OCSP cache
+	ocspCache := ocsp.NewCache()
+	connOpts = append(
+		connOpts,
+		WithOCSPCache(func(ocsp.Cache) ocsp.Cache { return ocspCache }),
+	)
+
+	// Disable communication with external OCSP responders.
+	if co.DisableOCSPEndpointCheck != nil {
+		connOpts = append(
+			connOpts,
+			WithDisableOCSPEndpointCheck(func(bool) bool { return *co.DisableOCSPEndpointCheck }),
+		)
 	}
 
-	cert, err := tls.X509KeyPair(certBlock, keyBlock)
-	if err != nil {
-		return "", err
+	// LoadBalanced
+	if co.LoadBalanced != nil {
+		cfgp.LoadBalanced = *co.LoadBalanced
+
+		serverOpts = append(
+			serverOpts,
+			WithServerLoadBalanced(func(bool) bool { return *co.LoadBalanced }),
+		)
+		connOpts = append(
+			connOpts,
+			WithConnectionLoadBalanced(func(bool) bool { return *co.LoadBalanced }),
+		)
 	}
 
-	cfg.Certificates = append(cfg.Certificates, cert)
+	serverOpts = append(
+		serverOpts,
+		WithClock(func(*session.ClusterClock) *session.ClusterClock { return clock }),
+		WithConnectionOptions(func(...ConnectionOption) []ConnectionOption { return connOpts }))
 
-	// The documentation for the tls.X509KeyPair indicates that the Leaf certificate is not
-	// retained.
-	crt, err := x509.ParseCertificate(certDecodedBlock)
-	if err != nil {
-		return "", err
+	cfgp.ServerOpts = serverOpts
+
+	// Deployment
+	if co.Deployment != nil {
+		// A deployment cannot be created with server or topology options. To extend the optional requirements
+		// add topology options to the "hasOptionalData" function.
+		if len(serverOpts) > 2+defaultOptions || hasOptionalData(*co) {
```

Comment:
We're not actually doing anything with the custom deployment here, so the custom deployment check seems out-of-place. We should keep the custom deployment check logic in `mongo.NewClient`.

Another option is to remove the "custom deployment with server/topology options" check. This is intended as a "don't shoot yourself in the foot" check, but has historically been buggy and using a custom deployment is already explicitly recommended against. Should we replace this check with a note that says "Server and topology options are not honored if a custom deployment is used"?

Reply:
I think if we have concerns about keeping `hasOptionalData` in sync with the logic, then it's probably best to just remove this branch.

Reply:
That sounds good to me. @benjirewis do you have any reservations about removing the "custom deployment with server/topology options" check?

The change that originally introduced that logic is https://github.com/mongodb/mongo-go-driver/commit/bb9530df7191054d68c7879dc4a2045eb71ad856. I don't see anything in the associated Jira ticket that specifically requires that check, so it seems to be a "nice to have".

Reply:
Great point; I think it's fine to remove as long as we add in a comment like what you suggested @matthewdale ("Server and topology options are not honored if a custom deployment is used").